docs/compatibility_draft.md

```diff
@@ -0,0 +1,246 @@
+
+|  R#   | Description | Priority |
+|:-----:|-------------|:--------:|
+|   1   | Must provide a minimum of 6 months backwards compatibility for serialized StableHLO programs. | Must have |
+|   2   | Must provide a minimum of 3 weeks forwards compatibility for serialized StableHLO programs. | Must have |
+|   3   | C++ and Python APIs should provide some degree of source compatibility. | Nice to have |
+
+## Compatibility Guarantees
+**Phase 1: Static Guarantees**
+
+In the short term StableHLO will provide static guarantees on stability. This includes IR structure, verifiers, and type inference. This will be expanded to include runtime tests as the interpreter becomes available. 
+
+*Backward compatibility:* After upgrade, the v1 file must match the semantics of the v2 file.
+
+*Forward compatibility:* After downgrade, the v2 file must match the semantics of the v1 file.
+
+**Phase 2: Dynamic Guarantees**
+
+A long term goal of this project is to verify result parity using the reference StableHLO interpreter.
+
+*Backward compatibility:* After upgrade, the v1 file must produce the same results or errors as the v2 file.
+
+*Forward compatibility:* After downgrade, the v2 file must produce the same results or errors as the v1 file.
+
+## Supported Dialects
+
+### Operations
+
+The following list of ops are use-case driven and can grow as needed. Supported operations will be validated using a pass before bytecoding. The list of ops and dialects to initially support is as follows:
+
+|  Dialect  | Ops |
+| --------- | --- |
+| Stablehlo | All ops except `CustomCallOp`. |
+| Builtin   | `ModuleOp` |
+| Func      | `FuncOp, ReturnOp` |
+| Arith     | `AddIOp, ArithmeticDialect, CmpIOp, CmpIPredicate, ConstantOp, DivSIOp, ExtSIOp, IndexCastOp, MaxSIOp, MinSIOp, MulIOp, SelectOp, SubIOp, TruncIOp` |
+| Tensor    | `DimOp, FromElementsOp` |
+
+### Attributes and Types
+All attributes and types used in StableHLO and dependent dialect ops are supported.
+
+## Upgrade / Downgrade Machinery
+There is currently no versioning for MLIR bytecode, however the proposals for maintaining compatibility described in the following section require the following:
+- Pre-serialization downgrade on generic operation form
+- Post-deserialization upgrade of generic operation form
+
+If the above machinery is not possible in a reasonable timeframe, many of these designs can fall back to using the _“Versioned Op Protocol”_ described below. 
+
+## Compatibility Protocols
+
+### Compatible Changes
+_Characteristics of a compatible change:_
+
+- Does not break forward or backward compatibility.
+- May require updating Python/C APIs.
+
+_Example:_ Rename an Operation, but no the mnemonic `SinOp → SineOp`:
+
+_Compatibility Protocol:_
+
+<img width="859" alt="compat" src="https://user-images.githubusercontent.com/3036970/189246138-12eaf8e3-5702-4de7-a1df-1e523f703d1e.png">
```

Comment:
How would you feel about this vs speaking more generically about the example? 

I.e. _Example:_ Rename an operation, but not the mnemonic `MyOp → TheOp`:

Part of me wanted to keep the diagrams as dialect agnostic as possible, since we are talking about a generic mechanism to keep compatibility. If we wanted to be specific and keep the example, it would be `MyDialect.MyOp {my_attr = 1}  → stablehlo.sine_op`